### PR TITLE
Add support for Neuron and HabanaLabs devices

### PR DESCRIPTION
**Pull request by mrgolin**
This PR adds abstraction layer for supporting various AI HW accelerators.

Add support for Neuron accelerator device memory usage.
To support Neuron, configure using  --enable-neuron --with-neuron=<path/to/neuron/base>.
Run bandwidth tests with Neuron direct by specifying --use_neuron <logical neuron core id> option.

Add support for Habana Labs accelerators device memory usage.
To support Habana Labs, configure using --enable-hl --with-hl=<habanalabs/installation/prefix>
Run bandwidth tests with Habana Labs direct by specifying --use_hl <pci_bus_id>.

### CODE REVIEW

**Comment by HassanKhadour** [comment]
Hi @mrgolin , thanks for the pull request!
can you please add informative commit message ?

**Comment by mrgolin** [comment]
@HassanKhadour Thanks for taking a look!
I've updated the commit message to better explain the change and its use case.

File: configure.ac
Comment (rajachan) [inline configure.ac:257]: Can we do this the proper way and introduce a configure flag with AC_ARG_WITH instead? The existing `CUDA_H_PATH` method in itself is not clean, we shouldn't add more such options.

**Review by rajachan — CHANGES_REQUESTED** [review]

File: configure.ac
Comment (rajachan) [reply configure.ac:257]: FWIW, I have a PR to fix this for CUDA https://github.com/linux-rdma/perftest/pull/152

File: src/perftest_parameters.h
Comment (galpress) [inline src/perftest_parameters.h:75]: Do you really want perftest to rely on experimental APIs?
It's probably wise to surround it with a dedicated ifdef..

**Comment by galpress** [comment]
Copy-pasting the same concepts every time a new accelerator is introduced isn't very scalable, I would argue it's time to abstract these and have specific "hardware providers" implementations.

**Comment by rajachan** [comment]
>  Copy-pasting the same concepts every time a new accelerator is introduced isn't very scalable, I would argue it's time to abstract these and have specific "hardware providers" implementations.

+1, abstracting out the accelerator alloc/copy routines would make things maintainable for the individual benchmarks going forward.

File: configure.ac
Comment (mrgolin) [reply configure.ac]: Thanks, changed to using AC_ARG_WITH as in your PR.

File: src/perftest_parameters.h
Comment (mrgolin) [reply src/perftest_parameters.h]: Removed this include entirely.

**Comment by mrgolin** [comment]
@galpress @rajachan Agree it's about time for refactoring HW accelerators memory usage code. I'll take this on.

**Comment by HassanKhadour** [comment]
Can one of the admins verify this patch?

**Comment by mrgolin** [comment]
I've added abstraction for various memory types as well as support for Habana Labs AI accelerators.
@galpress @rajachan @HassanKhadour I would appreciate your review.

**Comment by mrgolin** [comment]
@HassanKhadour Can you please share failing check results?

**Comment by HassanKhadour** [comment]
Thanks Michael for your commit, Will try to review ASAP.
following the cppcheck failures:

Perftest_memory

**Comment by HassanKhadour** [comment]
Hi Michael,
following the failure:

src/host_memory.c:68:26: style: Unused variable: host_ctx [unusedVariable]
12:14:56  struct host_memory_ctx *host_ctx;

**Comment by HassanKhadour** [comment]
The check has been passed, reported wrong status, fixed it

**Comment by mrgolin** [comment]
@galpress @rajachan @HassanKhadour any comments will be appreciated.

**Comment by mrgolin** [comment]
How is that this PR is still waiting for review after we've been asked to refactor perftest accelerators memory support and at the same time commits are pushed directly to master adding a new memory type (Cuda dma-buf) in an old way, blowing the code, making it unmaintainable (exactly as @gal-pressman and @rajachan mentioned) and making this PR unmergeable?

https://github.com/linux-rdma/perftest/commit/8b666479b6b6a18c1e9b9d6b8b0d3e82e0f677a1
https://github.com/linux-rdma/perftest/commit/e589f4cc658611796ef491d326ab1faa70d14315
https://github.com/linux-rdma/perftest/commit/b0823b7000c9130396a07749d692b730120dcb82
https://github.com/linux-rdma/perftest/commit/31159151cf90737ddd31b49fd1c412016a427978

**Comment by HassanKhadour** [comment]
Hi mrgolin,

We are sorry we weren't in the loop of the review from the beginning, we apologize about the inconvenience that is caused.
I want to ask @gal-pressman @rajachan who were in the loop if it would be okay if they rereview, I don't have any problem with merging this feature more than running few basic test in order to test its performance although I don't see any code that can possibly cause performance difference.
once having the needed review we can merge it and the dma_buf will be handled by me and @sshaulnv

**Comment by gal-pressman** [comment]
I merely suggested it doesn't make sense that the same code is copy-pasted again and again.
Since you're already pushing other code that does it, might as well merge this one as well..

**Comment by mrgolin** [comment]
> I merely suggested it doesn't make sense that the same code is copy-pasted again and again. Since you're already pushing other code that does it, might as well merge this one as well..

Gal, I did what you suggested and r

> I merely suggested it doesn't make sense that the same code is copy-pasted again and again. Since you're already pushing other code that does it, might as well merge this one as well..

Gal, I actually did what you suggested (See https://github.com/linux-rdma/perftest/pull/155/commits/457fc467ec70a3c4680a1444f71fac48b0968aab) and will appreciate if you can review this again.

File: configure.ac
Comment (rajachan) [reply configure.ac]: The autoconfig updates look good to me.

File: src/memory.h
Comment (rajachan) [inline src/memory.h:19]: Why not keep the ctx struct separate from the ops with an ops struct that gets initialized for each accelerator type?

File: src/memory.h
Comment (rajachan) [inline src/memory.h:22]: You might also need a copy hook, to copy data to and from the device. I don't think the current benchmarks use it, but if you wanted to add GPU support to, say, ib_write_lat... we would need a copy from the device to the host on the target side to check completions.

**Review by rajachan — COMMENTED** [review]
I don't think I can review the Habana/Neuron bits. I do like the intent of the memory type abstraction commit, so  I'd be fine merging it if you get sign offs on the other two commits.

File: src/memory.h
Comment (mrgolin) [reply src/memory.h:19]: We can do this but ctx struct is specific for each memory type so in perftest_parameters.c we will need allocate and init both ctx and ops for every memory type, and this I would try to avoid. What advantage do you see in holding them separately?

File: src/cuda_memory.c
Comment (mrgolin) [reply src/cuda_memory.c:19]: That is from original code (I've tried really hard to keep same behavior).
We can improve this in a separate commit.

File: src/cuda_memory.c
Comment (mrgolin) [reply src/cuda_memory.c:39]: Just to avoid changing behavior in a refactoring commit.

File: src/memory.h
Comment (rajachan) [reply src/memory.h:22]: Ack,  on this and the other comments about avoiding functional changes in a commit meant to refactor code.

File: src/perftest_parameters.c
Comment (HassanKhadour) [inline src/perftest_parameters.c:2645]: Can you please add description for this if condition?

File: src/perftest_resources.c
Comment (HassanKhadour) [inline src/perftest_resources.c:1316]: Are u sure thats its making the exact behaviour for mmap?, before it did munmap for only the first qp, now for all

File: src/memory.h
Comment (HassanKhadour) [inline src/memory.h:13]: Please add description for it

File: src/host_memory.h
Comment (HassanKhadour) [inline src/host_memory.h:9]: Do we have stdeff .h in C89?

**Review by HassanKhadour — CHANGES_REQUESTED** [review]
Thanks for your contribution.
The rdma_cm is damaged, getting segmentation fault after running with it.
Can you please fix it and address the comments ?

File: src/perftest_resources.c
Comment (mrgolin) [reply src/perftest_resources.c:1561]: No. It is always preceded by ctx->is_contig_supported = FAILURE.
Assumed may be used someone for manual debug / activation so didn't remove this.

File: src/perftest_resources.c
Comment (mrgolin) [reply src/perftest_resources.c:1559]: No. In addition I didn't any flow getting here with is_contig_supported uninitialized.

File: src/rocm_memory.h
Comment (mrgolin) [reply src/rocm_memory.h:12]: Not sure I understand. Can you elaborate?

File: src/perftest_resources.c
Comment (mrgolin) [reply src/perftest_resources.c:1791]: This code overruns previously allocated ctx->buf (array of pointers to buffers) by an aligned buffer of single mr size (why?).
I'm not sure what this code meant to do but removing it shouldn't affect functionality. What do you think? Maybe there was some performance related reason?

File: src/perftest_resources.c
Comment (mrgolin) [reply src/perftest_resources.c:1316]: I think that current mmap implementation can handle only num_of_qps=1 or mr_per_qp=false cases, see line 190:
```
	ctx->buf[0] = mmap(NULL, size, PROT_WRITE | PROT_READ, MAP_SHARED, fd,
			offset);
```
So dereg_counter should be equal to 1 in that case:
```
	dereg_counter = (user_param->mr_per_qp) ? user_param->num_of_qps : 1;
```

File: src/host_memory.h
Comment (mrgolin) [reply src/host_memory.h:9]: Yes it's in ANSI C89 standard.

**Comment by mrgolin** [comment]
> Thanks for your contribution. The rdma_cm is damaged, getting segmentation fault after running with it. Can you please fix it and address the comments ?

Thanks for your review. We are currently not using rdma_cm, can you please share dmesg and any additional info that might help find the issue.

**Comment by HassanKhadour** [comment]
> > Thanks for your contribution. The rdma_cm is damaged, getting segmentation fault after running with it. Can you please fix it and address the comments ?
> 
> Thanks for your review. We are currently not using rdma_cm, can you please share dmesg and any additional info that might help find the issue.

The segfault occurs in Client side.
I'll explain the flow that causes the segfault - e.g. for ib_write_bw:
establish_connection (line 120) --> rdma_client_connect --> ctx_init --> ctx->memory->init(ctx->memory)

But the ctx->memory still isn't initiated in this stage.

File: src/host_memory.h
Comment (HassanKhadour) [reply src/host_memory.h:9]: Ack

File: src/rocm_memory.h
Comment (HassanKhadour) [reply src/rocm_memory.h:12]: Sorry, not needed

File: src/perftest_resources.c
Comment (HassanKhadour) [reply src/perftest_resources.c:1559]: Ack

File: src/perftest_resources.c
Comment (HassanKhadour) [reply src/perftest_resources.c:1561]: Ack

File: src/perftest_resources.c
Comment (HassanKhadour) [reply src/perftest_resources.c:1316]: I think that we must make it as the current implementation and refactor that after if needed

File: src/perftest_resources.c
Comment (HassanKhadour) [reply src/perftest_resources.c:1791]: I'm not sure why, but It may be some performance related reason as you said, can you please make at as the current implementation?

Comment (mrgolin) [reply src/perftest_resources.c:1851]: Restored this reallocation code.
I think we should consider removing this in a separate PR.

File: src/perftest_parameters.c
Comment (mrgolin) [reply src/perftest_parameters.c:2659]: Done.

File: src/perftest_resources.c
Comment (mrgolin) [reply src/perftest_resources.c:1318]: Checked this again and we are keeping same behavior as before for mmap.

File: src/memory.h
Comment (mrgolin) [reply src/memory.h:14]: Done.

**Comment by mrgolin** [comment]
> > > Thanks for your contribution. The rdma_cm is damaged, getting segmentation fault after running with it. Can you please fix it and address the comments ?
> > 
> > 
> > Thanks for your review. We are currently not using rdma_cm, can you please share dmesg and any additional info that might help find the issue.
> 
> The segfault occurs in Client side. I'll explain the flow that causes the segfault - e.g. for ib_write_bw: establish_connection (line 120) --> rdma_client_connect --> ctx_init --> ctx->memory->init(ctx->memory)
> 
> But the ctx->memory still isn't initiated in this stage.

Thanks for specifying the error flow. Fixed.

**Comment by mrgolin** [comment]
@HassanKhadour we've rebased this PR on Cuda dmabuf support changes and adopted it to our refactoring.

**Comment by mrgolin** [comment]
@HassanKhadour Can you please share failing checks?

**Comment by HassanKhadour** [comment]
image

**Comment by mrgolin** [comment]
>All checks have failed
@HassanKhadour please elaborate.

**Comment by HassanKhadour** [comment]
image

File: src/host_memory.c
Comment (HassanKhadour) [inline src/host_memory.c:35]: remove  huge_shmid from pingpong_context

File: src/perftest_parameters.c
Comment (HassanKhadour) [inline src/perftest_parameters.c:3294]: add check rocm_memory_supported()

File: src/perftest_resources.c
Comment (HassanKhadour) [reply src/perftest_resources.c:1318]: How? I don't see perftest limit the number of qp's in case of mmap

File: src/rocm_memory.h
Comment (HassanKhadour) [inline src/rocm_memory.h:29]: I think we should add print that tells that rocm is unsupported

**Comment by sshaulnv** [comment]
the use_cuda with rdma_cm is damaged, getting segmentation fault.

File: src/host_memory.c
Comment (mrgolin) [reply src/host_memory.c:43]: This was a bug checking if ctx->buf is -1 right after assigning into ctx->buf[qp_index]:
```
ctx->buf[qp_index] = (void *) shmat(ctx->huge_shmid, SHMAT_ADDR, SHMAT_FLAGS);
if (ctx->buf == (void *) -1) {
```
I usually attempt not fixing any existing issues during refactoring but it was hard to keep this one. We can just remove this check assuming that if ctx->buf == -1 we will anyway get page fault before reaching it.

**Comment by mrgolin** [comment]
> the use_cuda with rdma_cm is damaged, getting segmentation fault.

Thanks! I fixed a bug in perftest_communication, please report if there are other related issues.

**Comment by mrgolin** [comment]
@HassanKhadour If there are no additional comments can you please merge this PR?

**Comment by HassanKhadour** [comment]
Hi @mrgolin , I had some additional comments on the code.
Sure we are working on merging the change and we have set a plan for it, but as you know its a huge change that touch all perftest code and we are trying to test everything in it.
Thanks

**Comment by mrgolin** [comment]
> Hi @mrgolin , I had some additional comments on the code. Sure we are working on merging the change and we have set a plan for it, but as you know its a huge change that touch all perftest code and we are trying to test everything in it. Thanks

Thanks! Please keep me looped and let me know if I can help anyhow.

**Comment by HassanKhadour** [comment]
Hi @mrgolin, Appreciate your help in testing the ROCm changes on devices that support ROCm & other vendor devices that supports RDMA, from our side I'm testing on Nvidia devices.
Also can u please answer the comments above?

Thanks

**Comment by sshaulnv** [comment]
the use_cuda with rdma_cm, UD connection and bidirectional is damaged.
**cmd for example:** ./ib_send_bw --rdma_cm -c UD --bidirectional --use_cuda 0

**Comment by mrgolin** [comment]
> the use_cuda with rdma_cm, UD connection and bidirectional is damaged.
> cmd for example: ./ib_send_bw --rdma_cm -c UD --bidirectional --use_cuda 0

@sshaulnv Can you please share more details? If possible a backtrace or a coredump.

**Comment by mrgolin** [comment]
> Hi @mrgolin, Appreciate your help in testing the ROCm changes on devices that support ROCm & other vendor devices that supports RDMA, from our side I'm testing on Nvidia devices. Also can u please answer the comments above?
> 
> Thanks

Sorry for a delayed reply, we've tested with host memory, Habalabs, Neuron, Cuda w/ and w/o dmabuf.
Unfortunately I couldn't get access to AMD accelerators to fully test ROCm, but I did partial testing by replacing lib hip calls.

**Comment by sshaulnv** [comment]
> > the use_cuda with rdma_cm, UD connection and bidirectional is damaged.
> > cmd for example: ./ib_send_bw --rdma_cm -c UD --bidirectional --use_cuda 0
> 
> @sshaulnv Can you please share more details? If possible a backtrace or a coredump.

sure.
i just used the following commands:
**server**: ./ib_send_bw --rdma_cm -c UD --bidirectional --use_cuda 0
**client**: ./ib_send_bw --rdma_cm -c UD --bidirectional 11.237.149.43

**backtrace:**
Thread 1 "ib_send_bw" received signal SIGSEGV, Segmentation fault.
ibv_init_ah_from_wc (context=0x7ffff5303150, port_num=1 '\001', wc=wc@entry=0x7fffffffdba0, grh=0x7fffcb200000, ah_attr=ah_attr@entry=0x7fffffffdb50)
    at /usr/src/debug/rdma-core-59mlnx44-1.fc32.59056.x86_64/libibverbs/verbs.c:934
934             set_ah_attr_generic_fields(ah_attr, wc, grh, port_num);
(gdb) bt
0  ibv_init_ah_from_wc (context=0x7ffff5303150, port_num=1 '\001', wc=wc@entry=0x7fffffffdba0, grh=0x7fffcb200000, ah_attr=ah_attr@entry=0x7fffffffdb50)
    at /usr/src/debug/rdma-core-59mlnx44-1.fc32.59056.x86_64/libibverbs/verbs.c:934
1  0x00007ffff7dc31b8 in ibv_create_ah_from_wc (pd=0x447e80, wc=wc@entry=0x7fffffffdba0, grh=<optimized out>, port_num=<optimized out>)
    at /usr/src/debug/rdma-core-59mlnx44-1.fc32.59056.x86_64/libibverbs/verbs.c:960
2  0x000000000040a459 in rdma_cm_initialize_ud_connection_parameters (ctx=ctx@entry=0x7fffffffe150, user_param=user_param@entry=0x7fffffffe300)
    at src/perftest_communication.c:2115
3  0x000000000040aa4b in rdma_cm_connection_request_handler (ctx=0x7fffffffe150, user_param=0x7fffffffe300, cma_id=<optimized out>)
    at src/perftest_communication.c:2397
4  0x000000000040ad07 in rdma_cm_connect_events (ctx=0x7fffffffe150, ctx@entry=0x7fffffffde90, user_param=user_param@entry=0x7fffffffe300)
    at src/perftest_communication.c:2526
5  0x000000000040af2e in rdma_cm_server_connection (ctx=0x7fffffffde90, ctx@entry=0x7fffffffe150, user_param=user_param@entry=0x7fffffffe300, 
    hints=hints@entry=0x7fffffffdf40) at src/perftest_communication.c:2636
6  0x000000000040b353 in create_rdma_cm_connection (ctx=ctx@entry=0x7fffffffe150, user_param=user_param@entry=0x7fffffffe300, 
    comm=comm@entry=0x7fffffffe020, my_dest=my_dest@entry=0x4475e0, rem_dest=rem_dest@entry=0x447620) at src/perftest_communication.c:2761
7  0x00000000004053eb in main (argc=<optimized out>, argv=<optimized out>) at src/send_bw.c:268

**Comment by mrgolin** [comment]
> > > the use_cuda with rdma_cm, UD connection and bidirectional is damaged.
> > > cmd for example: ./ib_send_bw --rdma_cm -c UD --bidirectional --use_cuda 0
> > 
> > 
> > @sshaulnv Can you please share more details? If possible a backtrace or a coredump.
> 
> sure. i just used the following commands: **server**: ./ib_send_bw --rdma_cm -c UD --bidirectional --use_cuda 0 **client**: ./ib_send_bw --rdma_cm -c UD --bidirectional 11.237.149.43
> 
> **backtrace:** Thread 1 "ib_send_bw" received signal SIGSEGV, Segmentation fault. ibv_init_ah_from_wc (context=0x7ffff5303150, port_num=1 '\001', wc=wc@entry=0x7fffffffdba0, grh=0x7fffcb200000, ah_attr=ah_attr@entry=0x7fffffffdb50) at /usr/src/debug/rdma-core-59mlnx44-1.fc32.59056.x86_64/libibverbs/verbs.c:934 934 set_ah_attr_generic_fields(ah_attr, wc, grh, port_num); (gdb) bt 0 ibv_init_ah_from_wc (context=0x7ffff5303150, port_num=1 '\001', wc=wc@entry=0x7fffffffdba0, grh=0x7fffcb200000, ah_attr=ah_attr@entry=0x7fffffffdb50) at /usr/src/debug/rdma-core-59mlnx44-1.fc32.59056.x86_64/libibverbs/verbs.c:934 1 0x00007ffff7dc31b8 in ibv_create_ah_from_wc (pd=0x447e80, wc=wc@entry=0x7fffffffdba0, grh=, port_num=) at /usr/src/debug/rdma-core-59mlnx44-1.fc32.59056.x86_64/libibverbs/verbs.c:960 2 0x000000000040a459 in rdma_cm_initialize_ud_connection_parameters (ctx=ctx@entry=0x7fffffffe150, user_param=user_param@entry=0x7fffffffe300) at src/perftest_communication.c:2115 3 0x000000000040aa4b in rdma_cm_connection_request_handler (ctx=0x7fffffffe150, user_param=0x7fffffffe300, cma_id=) at src/perftest_communication.c:2397 4 0x000000000040ad07 in rdma_cm_connect_events (ctx=0x7fffffffe150, ctx@entry=0x7fffffffde90, user_param=user_param@entry=0x7fffffffe300) at src/perftest_communication.c:2526 5 0x000000000040af2e in rdma_cm_server_connection (ctx=0x7fffffffde90, ctx@entry=0x7fffffffe150, user_param=user_param@entry=0x7fffffffe300, hints=hints@entry=0x7fffffffdf40) at src/perftest_communication.c:2636 6 0x000000000040b353 in create_rdma_cm_connection (ctx=ctx@entry=0x7fffffffe150, user_param=user_param@entry=0x7fffffffe300, comm=comm@entry=0x7fffffffe020, my_dest=my_dest@entry=0x4475e0, rem_dest=rem_dest@entry=0x447620) at src/perftest_communication.c:2761 7 0x00000000004053eb in main (argc=, argv=) at src/send_bw.c:268

@sshaulnv you could simply mention that you have pushed a fix for this long existing issue (and this PR is not rebased on..) instead of letting me debug this and wonder how my unrelated changes could cause a segfault in this code.
(https://github.com/linux-rdma/perftest/commit/aeb15b18f7fa344df5667a758ff3c453dcd5afc8)

**Comment by sshaulnv** [comment]
> > > > the use_cuda with rdma_cm, UD connection and bidirectional is damaged.
> > > > cmd for example: ./ib_send_bw --rdma_cm -c UD --bidirectional --use_cuda 0
> > > 
> > > 
> > > @sshaulnv Can you please share more details? If possible a backtrace or a coredump.
> > 
> > 
> > sure. i just used the following commands: **server**: ./ib_send_bw --rdma_cm -c UD --bidirectional --use_cuda 0 **client**: ./ib_send_bw --rdma_cm -c UD --bidirectional 11.237.149.43
> > **backtrace:** Thread 1 "ib_send_bw" received signal SIGSEGV, Segmentation fault. ibv_init_ah_from_wc (context=0x7ffff5303150, port_num=1 '\001', wc=wc@entry=0x7fffffffdba0, grh=0x7fffcb200000, ah_attr=ah_attr@entry=0x7fffffffdb50) at /usr/src/debug/rdma-core-59mlnx44-1.fc32.59056.x86_64/libibverbs/verbs.c:934 934 set_ah_attr_generic_fields(ah_attr, wc, grh, port_num); (gdb) bt 0 ibv_init_ah_from_wc (context=0x7ffff5303150, port_num=1 '\001', wc=wc@entry=0x7fffffffdba0, grh=0x7fffcb200000, ah_attr=ah_attr@entry=0x7fffffffdb50) at /usr/src/debug/rdma-core-59mlnx44-1.fc32.59056.x86_64/libibverbs/verbs.c:934 1 0x00007ffff7dc31b8 in ibv_create_ah_from_wc (pd=0x447e80, wc=wc@entry=0x7fffffffdba0, grh=, port_num=) at /usr/src/debug/rdma-core-59mlnx44-1.fc32.59056.x86_64/libibverbs/verbs.c:960 2 0x000000000040a459 in rdma_cm_initialize_ud_connection_parameters (ctx=ctx@entry=0x7fffffffe150, user_param=user_param@entry=0x7fffffffe300) at src/perftest_communication.c:2115 3 0x000000000040aa4b in rdma_cm_connection_request_handler (ctx=0x7fffffffe150, user_param=0x7fffffffe300, cma_id=) at src/perftest_communication.c:2397 4 0x000000000040ad07 in rdma_cm_connect_events (ctx=0x7fffffffe150, ctx@entry=0x7fffffffde90, user_param=user_param@entry=0x7fffffffe300) at src/perftest_communication.c:2526 5 0x000000000040af2e in rdma_cm_server_connection (ctx=0x7fffffffde90, ctx@entry=0x7fffffffe150, user_param=user_param@entry=0x7fffffffe300, hints=hints@entry=0x7fffffffdf40) at src/perftest_communication.c:2636 6 0x000000000040b353 in create_rdma_cm_connection (ctx=ctx@entry=0x7fffffffe150, user_param=user_param@entry=0x7fffffffe300, comm=comm@entry=0x7fffffffe020, my_dest=my_dest@entry=0x4475e0, rem_dest=rem_dest@entry=0x447620) at src/perftest_communication.c:2761 7 0x00000000004053eb in main (argc=, argv=) at src/send_bw.c:268
> 
> @sshaulnv you could simply mention that you have pushed a fix for this long existing issue (and this PR is not rebased on..) instead of letting me debug this and wonder how my unrelated changes could cause a segfault in this code. ([aeb15b1](https://github.com/linux-rdma/perftest/commit/aeb15b18f7fa344df5667a758ff3c453dcd5afc8))

@alex--m sorry about that, I didn't remember making a fix for it.

**Comment by mrgolin** [comment]
Is there any additional blocker from merging this PR?

**Comment by HassanKhadour** [comment]
Hi @mrgolin 

Please rebase the code for the last time, we will give it the last testing cycle and merge it very soon.

Thanks

File: src/perftest_resources.c
Comment (sshaulnv) [inline src/perftest_resources.c:1283]: Maybe I missed it, but I can't find the memory release of 'dmabuf_offset'

**Comment by mrgolin** [comment]
> Please rebase the code for the last time, we will give it the last testing cycle and merge it very soon.

@HassanKhadour
Hi, any progress with this?

**Comment by HassanKhadour** [comment]
Hi @mrgolin, Please check @sshaulnv comment

**Comment by mrgolin** [comment]
> Hi @mrgolin, Please check @sshaulnv comment

AFAIU all comments were addressed.
Can we merge this?

**Comment by HassanKhadour** [comment]
There is this comment from @sshaulnv "Maybe I missed it, but I can't find the memory release of 'dmabuf_offset'"
image

**Comment by mrgolin** [comment]
> There is this comment from @sshaulnv "Maybe I missed it, but I can't find the memory release of 'dmabuf_offset'" image

Ohh.. sorry forgot to submit.

File: src/perftest_resources.c
Comment (sshaulnv) [reply src/perftest_resources.c:1283]: sure, thanks

**Comment by HassanKhadour** [comment]
@mrgolin Thanks for your contribution!

**Pull request merged**
